Validation/HLTrigger/src/HLTGenValHistCollFilter.cc

```diff
@@ -0,0 +1,211 @@
+#include "Validation/HLTrigger/interface/HLTGenValHistCollFilter.h"
+
+// constructor
+HLTGenValHistCollFilter::HLTGenValHistCollFilter(edm::ParameterSet filterCollConfig) {
+  objType_ = filterCollConfig.getParameter<std::string>("objType");
+  filter_ = filterCollConfig.getParameter<std::string>("filterName");
+  path_ = filterCollConfig.getParameter<std::string>("pathName");
+  hltProcessName_ = filterCollConfig.getParameter<std::string>("hltProcessName");
+  dR2limit_ = filterCollConfig.getParameter<double>("dR2limit");
+}
+
+edm::ParameterSetDescription HLTGenValHistCollFilter::makePSetDescription() {
+  edm::ParameterSetDescription desc;
+  desc.add<std::string>("objType", "");
+  desc.add<std::string>("hltProcessName", "HLT");
+  desc.add<double>("dR2limit", 0.1);
+  desc.add<std::string>("filterName", "");
+  desc.add<std::string>("pathName", "");
+  return desc;
+}
+
+// general hist booking function, receiving configurations for 1D and 2D hists and calling the respective functions
+void HLTGenValHistCollFilter::bookHists(DQMStore::IBooker& iBooker,
+                                        const std::vector<edm::ParameterSet>& histConfigs,
+                                        const std::vector<edm::ParameterSet>& histConfigs2D) {
+  for (const auto& histConfig : histConfigs)
+    book1D(iBooker, histConfig);
+  for (const auto& histConfig : histConfigs2D)
+    book2D(iBooker, histConfig);
+}
+
+// histogram filling routine
+void HLTGenValHistCollFilter::fillHists(const HLTGenValObject& obj, edm::Handle<trigger::TriggerEvent>& triggerEvent) {
+  // this handles the "before" step, denoted by a "dummy" filter called "beforeAnyFilter"
+  // the histogram is filled without any additional requirements for all objects
+  if (filter_ == "beforeAnyFilter") {
+    for (auto& hist : hists_)
+      hist->fill(obj);
+  } else {
+    // main filling code
+
+    // get filter object from filter name
+    edm::InputTag filterTag(filter_, "", hltProcessName_);
+    size_t filterIndex = triggerEvent->filterIndex(filterTag);
+
+    // get trigger objects passing filter in question
+    trigger::TriggerObjectCollection allTriggerObjects = triggerEvent->getObjects();  // all objects
+    trigger::TriggerObjectCollection selectedObjects;  // vector to fill with objects passing our filter
+    if (filterIndex < triggerEvent->sizeFilters()) {
+      const auto& keys = triggerEvent->filterKeys(filterIndex);
+      for (unsigned short key : keys) {
+        trigger::TriggerObject foundObject = allTriggerObjects[key];
+        selectedObjects.push_back(foundObject);
+      }
+    }
+
+    // differentiate between event level and particle level variables
+    std::vector<std::string> eventLevelVariables = {"AK4HT", "AK8HT", "MET"};
```

Comment:
it would probably be better to have this as const static member variable to avoid it being reallocated over and over. Minor I know.  Not sure if compiler is smart enough to optimise this away. At the very least should be const.